proposals/init.md

```diff
@@ -0,0 +1,581 @@
+the bigger the cost of this boiler plate:
+
+```cs
+struct Point
+{
+    public int X { get; }
+    public int Y { get; }
+
+    public Point(int X, int Y)
+    {
+        this.X = x;
+        this.Y = y;
+    }
+}
+```
+
+The `init` accessor makes immutable objects more flexible by allowing the
+caller to mutate the members during the act of construction. That means the
+object's immutable properties can participate in object initializers and thus
+removes the need for all constructor boilerplate in the type. The `Point`
+type is now simply:
+
+```cs
+struct Point
+{
+    public int X { get; init; }
+    public int Y { get; init; }
+}
+```
+
+The consumer can then use object initializers to create the object
+
+```cs
+var p = new Point() { X = 42, Y = 13 };
+```
+
+## Detailed Design
+
+### init members
+An init only property is declared by using the `init` accessor in place of the 
+`set` accessor:
+
+```cs
+class Student
+{
+    public string FirstName { get; init; }
+    public string LastName { get; init; }
+}
+```
+
+An instance property containing an `init` accessor is considered settable in
+the following circumstances:
+
+- During an object initializer
+- Inside an instance constructor of the containing or derived type, on `this` or `base`
+- Inside the `init` accessor of any property, on `this` or `base`
+
+The times above in which the `init` accessors are settable are collectively
+referred to in this document as the construction phase of the object.
+
+This means the `Student` class can be used in the following ways:
+
+```cs
+var s = new Student()
+{
+    FirstName = "Jared",
+    LastName = "Parosns",
+};
+s.LastName = "Parsons"; // Error: LastName is not settable
+```
+
+The rules around when `init` accessors are settable extend across type
+hierarchies. If the member is accessible and the object is known to be in the
+construction phase then the member is settable. That specifically allows for
+the following:
+
+```cs
+class Base
+{
+    public bool Value { get; init; }
+}
+
+class Derived : Base
+{
+    Derived()
+    {
+        // Not allowed with get only properties but allowed with init
+        Value = true;
+    }
+}
+
+class Consumption
+{
+    void Example()
+    {
+        var d = new Derived() { Value = true; };
+    }
+}
+
+```
+
+At the point a `init` accessor is invoked the instance is known to be 
+in the open construction phase. Hence an `init` accessor is allowed to take 
+the following actions in addition to what a normal `set` accessor can do:
+
+1. Call other `init` accessors available through `this` or `base`
+1. Assign `readonly` fields declared on the same type
+
+```cs
+class Complex
+{
+    readonly int Field1;
+    int Field2;
+    int Prop1 { get; init ; }
+    int Prop2
+    {
+        get => 42;
+        init
+        {
+            Field1 = 13; // okay
+            Field2 = 13; // okay
+            Prop1 = 13; // okay
+        }
```

Comment:
Why do you want to add that, and what would be the advantage? To me, this is duplicating the job of a constructor, and I think this will only create less readable code. Maybe there is a use-case that I missed; if so, feel free to enlighten me.

Reply:
I believe the motivation section of this proposal adequately outlines our reasoning: we want to support nominal, immutable construction patterns which we do not do today.

Reply:
> I believe the motivation section of this proposal adequately outlines our reasoning: we want to support nominal, immutable construction patterns which we do not do today.

Yes, I get the intent @333fred , which is a really good idea; less boilerplate is always welcome. I still remember the day you added auto-implemented properties; what a day!

But what is the difference between the following two blocks of code?

```csharp
init
{
    Field1 = 13; // okay
    Field2 = 13; // okay
    Prop1 = 13; // okay
}
```

```csharp
public SomeCtor()
{
    Field1 = 13; // okay
    Field2 = 13; // okay
    Prop1 = 13; // okay
}
```

Then what is the added value of that new way to initialize fields/props? (for C# users)

An intend is one thing, but does that help solve a problem or does it only create noise and possible confusion? Think about it, people will have one more place to put initialization logic in, does someone need that? If yes, who does? What use-case does this solve that was unsolvable before or that required more boilerplate?

Reply:
Think about optional parameters in a constructor today: you have a location that can default them if not provided, or throw if a combination of provided parameters is invalid. This adds the same logic for properties that are inited by the caller, but then immutable afterwards.

Reply:
> Think about optional parameters in a constructor today: you have a location that can default them if not provided, or throw if a combination of provided parameters is invalid. This adds the same logic for properties that are inited by the caller, but then immutable afterwards.

When I think about that I see types that could require some kind of possible value/props/null/not null combo but without any way to know it in advance from the API of that class/struct. People can be creative 😉 

Then what about I add another property that initializes `init` and `readonly` props/fields as well?

```cs
class Complex
{
    readonly int Field1;
    int Field2;
    int Prop1 { get; init ; }

    int Prop2
    {
        get => 42;
        init
        {
            Field1 = 13; // okay
            Field2 = 13; // okay
            Prop1 = 13; // okay
        }
    }

    int Prop3
    {
        get => 123;
        init
        {
            Field1 = 3;
            Field2 = 3;
            Prop1 = 3;
        }
    }
}
```

In what order are those going to be initialized?

Wouldn't it be clearer to do the following?

```cs
class Complex
{
    readonly int Field1;
    int Field2;
    int Prop1 { get; init ; }
    int Prop2 { get; init; }
    int Prop3 { get; init; }

    public Complex()
    {
        Field1 = 13;
        Field2 = 13;
        Prop1 = 13;
        Prop2 = 42;
        Prop3 = 123;
    }
}
```

Reply:
You shouldn't necessarily take syntax examples from a proposal that merely illustrate what things are legal where as a good API design. More than likely, most `init` properties would initialize one, and only one, readonly field. You can still write your example today, using `set` instead of `init`, and it provides a markedly worse experience if you wanted it to actually be an immutable object.

As to ordering, that's the thing here: this is about _nominal_ object creation. Nominal creation is unordered, that's the whole point. If you want to control the order in which properties are set, then you need to use the existing tools we have today in positional creation (namely, standard constructor arguments).

Imagine you had an example like this:

```cs
public class PathObject
{
    string ActualPath { get; init; }
    bool SupportsLongPath { get; init; }
    init
    {
        if (ActualPath == null || !supportsLongPath && ActualPath.Length >= 256)
        {
            throw new ArgumentException(nameof(ActualPath), $"Long path support not enabled, but path is {ActualPath.Length}!");
        }
    }
}
```

Imagine if you had to write that without `init` support? You literally can't do it. There's no way today to ensure that `ActualPath` was initialized by the caller. Further, for validation of the length of `ActualPath`, you can't do that either: you need to ensure that `SupportsLongPath` is set _before_ attempting to validate the path length. The only way to handle that correctly would be to use a constructor, which guarantees ordering. However, that brings burdens of its own. Let's say that this is a public class in a library that you publish publicly. You decide that you want to add support for hidden path object creation. But there's a problem: you've already published this class. If you want to avoid breaking your customers, you're going to have to maintain this 2-parameter constructor forever more, because removing it would break existing customer binaries. In the `init` world, this pain point is gone. You add a new init-only property, validate as necessary in the `init` body, and no one who compiled against a previous version of your library has to care about this new optional parameter that you've added. For a very real example of this type of pain, look at https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/CSharpCompilationOptions.cs#L787-L913. We cannot remove these constructors, as existing binaries depend on them, even the constructor marked as `Bad - DO NOT USE`. Those constructors will never be able to go away, unfortunately, but at least we won't have to add more of them.

Reply:
Good point, I could write my messy code in a `set`, so `init` is similar in that way (well, its an "initializable `readonly` property setter" after all).

That said, I did not see the "init constructor-like" that you used in your last sample anywhere in the proposal, maybe you should add that as it seems that will be needed.

If the initialization is unordered, how can you be sure that your constructor-like-init code block will be hit **after** the init setter of the other properties?

More on that, to make the usage clearer for users, maybe `afterinit` would be easier to understand for non-compiler/csharplang-guru? It would also point people in the direction of centralizing initialization logic instead of splitting it everywhere. Maybe a `beforeinit` could benefit centralizing initialization while `afterinit` benefits validation?

Or to make it less "magic" and more discoverable, what about an `IInitializable` interface that adds a `void AfterInit();` and a `void BeforeInit()` methods usable during the initialization process? Same idea here, you need additional logic, you implement the interface (or create the `afterinit` block) and code your initialization logic there; otherwise, you don't pollute your code with it.

_As I understand it, you seem to want to help people get into C#, so making usage and understanding of new (and existing) features as easy as possible for everyone would probably help that. Just for fun, ask people in what order are executed auto-property initializers and constructors to see how unclear it can be. With `init` you will be adding another layer of "confusion" over that._

As for breaking changes, a new major version should allow you to clean stuff up and introduce breaking changes. Everyone makes mistakes but having to carry all of the mistakes of everyone forever ain't helping anyone 😉

Reply:
> That said, I did not see the "init constructor-like" that you used in your last sample anywhere in the proposal, maybe you should add that as it seems that will be needed.

It'll be coming in a separate proposal. It'll all part of a greater feature unity we're aiming for as part of records.

> If the initialization is unordered, how can you be sure that your constructor-like-init code block will be hit after the init setter of the other properties?

The compiler will be required to generate a call to the validator (straw-man name we're using for now) after an object initializer is run, so ordering won't matter.

Don't take the syntax I used as "How it will look exactly". A future proposal with more validator design will be coming, and we haven't talked much about the specific syntax beyond "We like the general idea".

Reply:
Good to know, thanks, I'll try to follow that up then.